File: server/internal/biz/customer_config_transition_test.go
package biz

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"
)

func TestCustomerConfigTransitionCheckAllowsUnchangedPublishedRevision(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)

	activeInput := validCustomerConfigInput()
	activeInput.Revision = "rev-a"
	active, err := uc.PublishCustomerConfig(ctx, activeInput, 10)
	if err != nil {
		t.Fatalf("publish active: %v", err)
	}
	if _, err := uc.ActivateCustomerConfig(ctx, active.CustomerKey, active.Revision, active.ConfigHash, active.ProductVersion, "", 10); err != nil {
		t.Fatalf("activate rev-a: %v", err)
	}
	targetInput := validCustomerConfigInput()
	targetInput.Revision = "rev-b"
	target, err := uc.PublishCustomerConfig(ctx, targetInput, 11)
	if err != nil {
		t.Fatalf("publish target: %v", err)
	}

	check, err := uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionActivate,
		CustomerKey:            target.CustomerKey,
		TargetRevision:         target.Revision,
		ExpectedConfigHash:     target.ConfigHash,
		ExpectedProductVersion: target.ProductVersion,
		ExpectedActiveRevision: active.Revision,
	})
	if err != nil {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
	if !check.Allowed || check.Noop || len(check.Blockers) != 0 {
		t.Fatalf("check = %#v", check)
	}
}

func TestCustomerConfigTransitionCheckBlocksRuntimeBreakingChanges(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)

	activeInput := validCustomerConfigInput()
	activeInput.Revision = "rev-a"
	addRuntimeProcessSelection(
		&activeInput,
		ProcessKeySalesOrderAcceptance,
		"v1",
		CustomerProcessVariantSalesApprovalPMC,
		"sales_order",
	)
	active, err := uc.PublishCustomerConfig(ctx, activeInput, 10)
	if err != nil {
		t.Fatalf("publish active: %v", err)
	}
	if _, err := uc.ActivateCustomerConfig(ctx, active.CustomerKey, active.Revision, active.ConfigHash, active.ProductVersion, "", 10); err != nil {
		t.Fatalf("activate rev-a: %v", err)
	}

	targetInput := validCustomerConfigInput()
	targetInput.Revision = "rev-b"
	addRuntimeProcessSelection(
		&targetInput,
		ProcessKeySalesOrderAcceptance,
		"v1",
		CustomerProcessVariantSalesApprovalEngineeringPMC,
		"sales_order",
	)
	for index := range targetInput.ModuleStates {
		if targetInput.ModuleStates[index].ModuleKey == "production" {
			targetInput.ModuleStates[index].State = "disabled"
		}
	}
	for index := range targetInput.RoleProfiles {
		if targetInput.RoleProfiles[index].RoleKey == FinanceRoleKey {
			targetInput.RoleProfiles[index].Disabled = true
		}
	}
	target, err := uc.PublishCustomerConfig(ctx, targetInput, 11)
	if err != nil {
		t.Fatalf("publish target: %v", err)
	}
	repo.processCount[customerRevisionKey(active.CustomerKey, active.Revision)+"/"+ProcessKeySalesOrderAcceptance] = 2
	repo.taskCount[customerRevisionKey(active.CustomerKey, active.Revision)+"/sales"] = 3
	repo.businessCount[active.CustomerKey+"/production"] = 4

	check, err := uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionActivate,
		CustomerKey:            target.CustomerKey,
		TargetRevision:         target.Revision,
		ExpectedConfigHash:     target.ConfigHash,
		ExpectedProductVersion: target.ProductVersion,
		ExpectedActiveRevision: active.Revision,
	})
	if err != nil {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
	if check.Allowed {
		t.Fatalf("breaking transition must be blocked: %#v", check)
	}
	assertCustomerConfigTransitionBlocker(t, check, "in_flight_processes_for_changed_contracts", 2)
	assertCustomerConfigTransitionBlocker(t, check, "open_workflow_tasks_for_changed_responsibility", 3)
	assertCustomerConfigTransitionBlocker(t, check, "open_business_documents_for_disabled_modules", 4)
}

func TestCustomerConfigTransitionCheckRejectsStaleIdentityAndUnactivatedRollback(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)

	activeInput := validCustomerConfigInput()
	activeInput.Revision = "rev-a"
	active, err := uc.PublishCustomerConfig(ctx, activeInput, 10)
	if err != nil {
		t.Fatalf("publish active: %v", err)
	}
	if _, err := uc.ActivateCustomerConfig(ctx, active.CustomerKey, active.Revision, active.ConfigHash, active.ProductVersion, "", 10); err != nil {
		t.Fatalf("activate rev-a: %v", err)
	}
	targetInput := validCustomerConfigInput()
	targetInput.Revision = "rev-b"
	target, err := uc.PublishCustomerConfig(ctx, targetInput, 11)
	if err != nil {
		t.Fatalf("publish target: %v", err)
	}

	check, err := uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionRollback,
		CustomerKey:            target.CustomerKey,
		TargetRevision:         target.Revision,
		ExpectedConfigHash:     strings.Repeat("a", 64),
		ExpectedProductVersion: "stale-product",
		ExpectedActiveRevision: "stale-revision",
	})
	if err != nil {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
	for _, code := range []string{
		"active_revision_changed",
		"rollback_target_not_activated",
		"target_config_hash_mismatch",
		"target_product_version_mismatch",
	} {
		assertCustomerConfigTransitionBlocker(t, check, code, 0)
	}
}

func TestCustomerConfigTransitionCheckRollbackRequiresImmutableActivatedHistory(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)
	now := time.Now()
	repo.revisions[customerRevisionKey("yoyoosun", "rev-old")] = &CustomerConfigRevision{
		CustomerKey:       "yoyoosun",
		Revision:          "rev-old",
		ProductVersion:    "product-v1",
		ConfigHash:        strings.Repeat("b", 64),
		ConfigHashVersion: CustomerConfigHashVersion,
		Status:            CustomerConfigStatusSuperseded,
		ActivatedAt:       &now,
	}
	repo.revisions[customerRevisionKey("yoyoosun", "rev-active")] = &CustomerConfigRevision{
		CustomerKey:       "yoyoosun",
		Revision:          "rev-active",
		ProductVersion:    "product-v1",
		ConfigHash:        strings.Repeat("c", 64),
		ConfigHashVersion: CustomerConfigHashVersion,
		Status:            CustomerConfigStatusActive,
		ActivatedAt:       &now,
	}

	check, err := uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionRollback,
		CustomerKey:            "yoyoosun",
		TargetRevision:         "rev-old",
		ExpectedConfigHash:     strings.Repeat("b", 64),
		ExpectedProductVersion: "product-v1",
		ExpectedActiveRevision: "rev-active",
	})
	if err != nil {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
	if !check.Allowed {
		t.Fatalf("activated immutable history should be a valid rollback target: %#v", check)
	}
}

func TestCustomerConfigTransitionCheckRejectsRollbackWithoutActiveRevision(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)
	now := time.Now()
	repo.revisions[customerRevisionKey("yoyoosun", "rev-old")] = &CustomerConfigRevision{
		CustomerKey:       "yoyoosun",
		Revision:          "rev-old",
		ProductVersion:    "product-v1",
		ConfigHash:        strings.Repeat("b", 64),
		ConfigHashVersion: CustomerConfigHashVersion,
		Status:            CustomerConfigStatusSuperseded,
		ActivatedAt:       &now,
	}

	check, err := uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionRollback,
		CustomerKey:            "yoyoosun",
		TargetRevision:         "rev-old",
		ExpectedConfigHash:     strings.Repeat("b", 64),
		ExpectedProductVersion: "product-v1",
		ExpectedActiveRevision: "",
	})
	if err != nil {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
	if check.Allowed {
		t.Fatalf("rollback without an active revision must fail closed: %#v", check)
	}
	assertCustomerConfigTransitionBlocker(t, check, "rollback_active_revision_required", 0)
}

func TestCustomerConfigTransitionCheckRejectsCurrentActiveAsRollbackTarget(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)
	now := time.Now()
	repo.revisions[customerRevisionKey("yoyoosun", "rev-active")] = &CustomerConfigRevision{
		CustomerKey:       "yoyoosun",
		Revision:          "rev-active",
		ProductVersion:    "product-v1",
		ConfigHash:        strings.Repeat("c", 64),
		ConfigHashVersion: CustomerConfigHashVersion,
		Status:            CustomerConfigStatusActive,
		ActivatedAt:       &now,
	}

	check, err := uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionRollback,
		CustomerKey:            "yoyoosun",
		TargetRevision:         "rev-active",
		ExpectedConfigHash:     strings.Repeat("c", 64),
		ExpectedProductVersion: "product-v1",
		ExpectedActiveRevision: "rev-active",
	})
	if err != nil {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
	if check.Allowed || check.Noop {
		t.Fatalf("current active revision must not be accepted as rollback target: %#v", check)
	}
	assertCustomerConfigTransitionBlocker(t, check, "target_status_invalid", 0)
}

func TestCustomerConfigTransitionCheckRejectsSupersededActivateAndUnknownHashVersion(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)
	now := time.Now()
	repo.revisions[customerRevisionKey("yoyoosun", "rev-old")] = &CustomerConfigRevision{
		CustomerKey:       "yoyoosun",
		Revision:          "rev-old",
		ProductVersion:    "product-v1",
		ConfigHash:        strings.Repeat("b", 64),
		ConfigHashVersion: CustomerConfigHashVersion + 1,
		Status:            CustomerConfigStatusSuperseded,
		ActivatedAt:       &now,
	}

	check, err := uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionActivate,
		CustomerKey:            "yoyoosun",
		TargetRevision:         "rev-old",
		ExpectedConfigHash:     strings.Repeat("b", 64),
		ExpectedProductVersion: "product-v1",
		ExpectedActiveRevision: "",
	})
	if err != nil {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
	if check.Allowed {
		t.Fatalf("superseded revision with an unknown hash version must not be activated: %#v", check)
	}
	assertCustomerConfigTransitionBlocker(t, check, "target_config_hash_version_invalid", 0)
	assertCustomerConfigTransitionBlocker(t, check, "target_status_invalid", 0)
}

func TestCustomerConfigTransitionCheckFailsClosedOnRepoError(t *testing.T) {
	ctx := context.Background()
	repo := newMemCustomerConfigRepo()
	uc := NewCustomerConfigUsecase(repo)
	in := validCustomerConfigInput()
	target, err := uc.PublishCustomerConfig(ctx, in, 10)
	if err != nil {
		t.Fatalf("publish target: %v", err)
	}
	repo.activeErr = errors.New("active lookup failed")
	_, err = uc.CheckCustomerConfigTransition(ctx, CustomerConfigTransitionCheckInput{
		Action:                 CustomerConfigTransitionActivate,
		CustomerKey:            target.CustomerKey,
		TargetRevision:         target.Revision,
		ExpectedConfigHash:     target.ConfigHash,
		ExpectedProductVersion: target.ProductVersion,
		ExpectedActiveRevision: "",
	})
	if err == nil || err.Error() != "active lookup failed" {
		t.Fatalf("CheckCustomerConfigTransition() error = %v", err)
	}
}

func assertCustomerConfigTransitionBlocker(t *testing.T, check *CustomerConfigTransitionCheck, code string, count int) {
	t.Helper()
	for _, blocker := range check.Blockers {
		if blocker.Code == code {
			if blocker.Count != count {
				t.Fatalf("blocker %s count = %d, want %d", code, blocker.Count, count)
			}
			return
		}
	}
	t.Fatalf("missing blocker %s in %#v", code, check.Blockers)
}

File: web/src/erp/api/customerConfigTransition.test.mjs
import assert from 'node:assert/strict'
import test from 'node:test'
import {
  CUSTOMER_CONFIG_HASH_VERSION,
  assertEffectiveCustomerConfigIdentity,
  assertPublishedCustomerConfigIdentity,
  buildCustomerConfigMutationPayload,
  confirmCustomerConfigTransition,
  resolveCustomerConfigApplyTransitionAction,
} from './customerConfigTransition.mjs'

const configHash = 'a'.repeat(64)
const manifest = {
  customer_key: 'yoyoosun',
  revision: 'rev-2',
  product_version: 'product-v1',
}
const validation = {
  customer_key: 'yoyoosun',
  revision: 'rev-2',
  config_hash: configHash,
  config_hash_version: CUSTOMER_CONFIG_HASH_VERSION,
  compiled_snapshot_ok: true,
}

function transition(params, overrides = {}) {
  return {
    action: params.action,
    customer_key: params.customer_key,
    target_revision: params.target_revision,
    target_config_hash: params.expected_config_hash,
    target_product_version: params.expected_product_version,
    expected_active_revision: params.expected_active_revision,
    observed_active_revision: '',
    allowed: true,
    noop: false,
    blockers: [],
    ...overrides,
  }
}

test('mutation payload 使用 activate/rollback 各自唯一字段合同', () => {
  assert.deepEqual(
    buildCustomerConfigMutationPayload('activate', {
      customer_key: 'yoyoosun',
      revision: 'rev-2',
      expected_config_hash: configHash,
      expected_product_version: 'product-v1',
      expected_active_revision: '',
    }),
    {
      customer_key: 'yoyoosun',
      revision: 'rev-2',
      expected_config_hash: configHash,
      expected_product_version: 'product-v1',
      expected_active_revision: '',
    }
  )
  const rollback = buildCustomerConfigMutationPayload('rollback', {
    customer_key: 'yoyoosun',
    target_revision: 'rev-1',
    expected_config_hash: configHash,
    expected_product_version: 'product-v1',
    expected_active_revision: 'rev-2',
  })
  assert.equal(rollback.target_revision, 'rev-1')
  assert.equal('revision' in rollback, false)
  assert.throws(
    () =>
      buildCustomerConfigMutationPayload('rollback', {
        customer_key: 'yoyoosun',
        revision: 'rev-1',
        expected_config_hash: configHash,
        expected_product_version: 'product-v1',
        expected_active_revision: 'rev-2',
      }),
    /未支持参数/
  )
  assert.throws(
    () =>
      buildCustomerConfigMutationPayload('activate', {
        customer_key: 'yoyoosun',
        revision: 'rev-2',
        expected_config_hash: configHash,
        expected_product_version: 'product-v1',
      }),
    /缺少当前有效配置版本/
  )
})

test('transition discovery 最多确认两次并复用同一 CAS identity', async () => {
  const calls = []
  const result = await confirmCustomerConfigTransition({
    action: 'activate',
    manifest,
    validation,
    check: async (params) => {
      calls.push(params)
      const confirmed = params.expected_active_revision === 'rev-1'
      return transition(params, {
        observed_active_revision: 'rev-1',
        allowed: confirmed,
        blockers: confirmed ? [] : [{ code: 'active_revision_changed' }],
      })
    },
  })

  assert.equal(calls.length, 2)
  assert.equal(calls[0].expected_active_revision, '')
  assert.equal(calls[1].expected_active_revision, 'rev-1')
  assert.equal(result.attempts, 2)
  assert.deepEqual(result.mutationPayload, {
    customer_key: 'yoyoosun',
    revision: 'rev-2',
    expected_config_hash: configHash,
    expected_product_version: 'product-v1',
    expected_active_revision: 'rev-1',
  })
})

test('transition blocker 和二次 active 漂移均 fail closed', async () => {
  await assert.rejects(
    () =>
      confirmCustomerConfigTransition({
        action: 'activate',
        manifest,
        validation,
        check: async (params) =>
          transition(params, {
            allowed: false,
            blockers: [{ code: 'target_module_closure_invalid' }],
          }),
      }),
    /不满足切换条件/
  )

  let callCount = 0
  await assert.rejects(
    () =>
      confirmCustomerConfigTransition({
        action: 'activate',
        manifest,
        validation,
        check: async (params) => {
          callCount += 1
          return transition(params, {
            expected_active_revision:
              callCount === 1 ? params.expected_active_revision : 'rev-1',
            observed_active_revision: callCount === 1 ? 'rev-1' : 'rev-3',
            allowed: callCount === 2,
          })
        },
      }),
    /发生变化/
  )
  assert.equal(callCount, 2)
})

test('publish 与 effective readback 必须匹配 canonical identity', () => {
  const published = {
    customer_key: 'yoyoosun',
    revision: 'rev-2',
    product_version: 'product-v1',
    config_hash: configHash,
    config_hash_version: CUSTOMER_CONFIG_HASH_VERSION,
    status: 'published',
  }
  assert.equal(
    assertPublishedCustomerConfigIdentity(published, manifest, validation)
      .configHash,
    configHash
  )
  assert.throws(
    () =>
      assertPublishedCustomerConfigIdentity(
        { ...published, config_hash: 'b'.repeat(64) },
        manifest,
        validation
      ),
    /发布结果与已校验/
  )
  assert.equal(
    assertEffectiveCustomerConfigIdentity(
      {
        customer: { key: 'yoyoosun' },
        configRevision: 'rev-2',
        configHash,
        configHashVersion: CUSTOMER_CONFIG_HASH_VERSION,
        source: 'active_customer_config_revision',
      },
      manifest,
      validation
    ).configRevision,
    'rev-2'
  )
})

test('内容寻址版本 A-B-A 回切时 superseded 版本必须走 rollback', async () => {
  assert.equal(
    resolveCustomerConfigApplyTransitionAction({ status: 'published' }),
    'activate'
  )
  assert.equal(
    resolveCustomerConfigApplyTransitionAction({ status: 'active' }),
    'activate'
  )
  assert.equal(
    resolveCustomerConfigApplyTransitionAction({ status: 'superseded' }),
    'rollback'
  )
  assert.throws(
    () => resolveCustomerConfigApplyTransitionAction({ status: 'draft' }),
    /状态不支持/
  )

  const result = await confirmCustomerConfigTransition({
    action: 'rollback',
    manifest,
    validation,
    check: async (params) => {
      const confirmed = params.expected_active_revision === 'rev-3'
      return transition(params, {
        observed_active_revision: 'rev-3',
        allowed: confirmed,
        blockers: confirmed ? [] : [{ code: 'active_revision_changed' }],
      })
    },
  })
  assert.deepEqual(result.mutationPayload, {
    customer_key: 'yoyoosun',
    target_revision: 'rev-2',
    expected_config_hash: configHash,
    expected_product_version: 'product-v1',
    expected_active_revision: 'rev-3',
  })
})

test('effective readback 的客户身份不一致时 fail closed', () => {
  assert.throws(
    () =>
      assertEffectiveCustomerConfigIdentity(
        {
          customer: { key: 'other-customer' },
          configRevision: 'rev-2',
          configHash,
          configHashVersion: CUSTOMER_CONFIG_HASH_VERSION,
          source: 'active_customer_config_revision',
        },
        manifest,
        validation
      ),
    /有效配置读回与已校验/
  )
})
